pkg/cmd/pulumi/destroy.go

```diff
@@ -222,8 +222,15 @@ func newDestroyCmd() *cobra.Command {
 				return result.FromError(err)
 			}
 
-			if len(*targets) > 0 && excludeProtected {
-				return result.FromError(errors.New("You cannot specify --target and --exclude-protected"))
+			if len(*targets) > 0 {
+				if excludeProtected {
+					return result.FromError(errors.New("You cannot specify --target and --exclude-protected"))
+				}
+				if snap != nil {
+					if err := checkTargets(*targets, snap); err != nil {
```

Comment:
Can this not still be in the deployment_executor? I'd think refresh ought to have the same check as well really (that's a change in behaviour but like destroy if you tell us to refresh X and it doesn't exist in the snapshot at the start then we know it's not gonna be created by refresh).
Maybe an option next to targets like `CheckTargets` or `TargetsMustExist`?

Reply:
This has given me the following idea that I'm going to pursue.

I'm going to add a field to the `type deployment struct` to contain a functions `DeploymentPreChecks` and `DeploymentPostChecks` that contains a struct with `Targets` and potentially other future things and pull this logic into `pkg/engine/deployment` which knows whether it's an `update`, `import`, `refresh`, `destroy`, etc.

Everything in `pkg/resource/deploy` is considered as a `deployment` and destroy/update/refresh logic is treated as part of `pkg/engine` and is leaking into it.

## Context

The `deployment_executor` doesn't know the difference between an `update`, `refresh` or a `destroy`.

Destroy and Refresh-ness of an update is based on the SourceFunc in the engine which is not sent to the deployment executor.

The deployment_executor used `DestroyTargets`, `RefreshTargets`, and `UpdateTargets` as a weak way to determine what the caller is doing to verify the legitimacy of the supplied targets.

Reply:
@Frassle on further inspection, this would be a bit more involved and I think we can start with a smaller and safer change.